### ISSUE TITLE
[Documentation] Update autocorrect docs to accurately define safety

### ISSUE THREAD

**Issue by a-lavis** [rubocop/rubocop#13951]
https://docs.rubocop.org/rubocop/usage/autocorrect.html

https://github.com/rubocop/rubocop/blob/0037549434928f40c0862dab1d15f88fe9e2977a/docs/modules/ROOT/pages/usage/autocorrect.adoc?plain=1#L17-L19

> - Some automatic corrections might change (slightly) the semantics of the code,
meaning they'd produce code that's mostly equivalent to the original code, but
not 100% equivalent. We call such autocorrect behavior "unsafe".

https://github.com/rubocop/rubocop/blob/0037549434928f40c0862dab1d15f88fe9e2977a/docs/modules/ROOT/pages/usage/autocorrect.adoc?plain=1#L32-L41

> In RuboCop 0.60, we began to annotate cops as `Safe` or not safe. The definition of
> safety is that the cop doesn't generate false positives. On top of that there's `SafeAutoCorrect`
> that might be set to `false` in cases where only the autocorrect performed by a cop
> is unsafe, but that the offense detection logic is safe. To sum it up:
>
> * Safe (`true/false`) - indicates whether the cop can yield false positives (by
design) or not.
> * SafeAutoCorrect (`true/false`) - indicates whether the autocorrect a cop
does is safe (equivalent) by design. If a cop is unsafe its autocorrect automatically
becomes unsafe as well.

According to this definition:
> The definition of safety is that the cop doesn't generate false positives.

`Style/RedundantCondition` would be considered unsafe, because it can generate false positives.  See #13931 for an example of the false positives it can generate.

It sounds like we want to keep this cop marked as safe (see discussion in #13931).  In that case, can we update the autocorrect docs to provide a new definition of safety?

I'm not sure what the new definition would be - maybe something along the lines of this comment: https://github.com/rubocop/rubocop/issues/13931#issuecomment-2690731329

**Comment by a-lavis** [comment]
Sorry about the `bug` label being added - this should probably be labelled with the `documentation` label instead.

**Comment by dvandersluis** [comment]
I don't think that the definition of `Safe` should be changed to suit a single cop. The linked issue doesn't read to me like it's settled that `Style/RedundantCondition` should not be marked unsafe, so that's the discussion to be had instead.

**Comment by a-lavis** [comment]
Ah okay yeah, that makes sense.  I can close this issue and ask for #13931 to be re-opened instead, would that be better?

**Comment by Earlopain** [comment]
This doesn't necessarily apply to `Style/RedundantCondition`, these are just my general thoughs on safety:

-------

Safety _is_ a bit more complicated. Very strictly speaking, `Lint/UselessAssignment` is unsafe because it changes the capture of a binding:

```rb
$binding = nil

def foo
  bar = 123 # Removing this breaks the last line
  $binding = binding
end

foo

puts $binding.local_variable_get(:bar)
```

This basic example doesn't make much sense but I've seen similar where such a binding was used in the context of erb execution. Removing the local variable has broken real code.

If you look hard enough, many currently safe cops can change something observable that a fraction of users may rely on. Simply being able redefine _any_ method throws a wrench in that.
